squad/core/models.py

```diff
@@ -563,6 +563,31 @@ def has_extra_metadata(self):
             return False
         return True
 
+    __failures__ = None
+
+    @property
+    def failures(self):
+        if self.__failures__ is None:
+            failures = []
+            limit = self.project.build_confidence_count
+
+            failing = self.tests.filter(result=False).exclude(has_known_issues=True).prefetch_related("metadata", "environment")
+
+            for f in failing:
+                failures.append(f)
+
+                history = Test.objects.filter(
+                    metadata=f.metadata,
+                    environment=f.environment,
+                    build_id__lt=self.id,
+                ).order_by("-build_id")[:limit]
```

Comment:
there is still one such query for each failing test. you want something like this, replacing the entire loop:

```python
# get all results for the currently failing tests in the last N builds, in a
# single query.
history = Test.objects.filter(
    metadata__in=[f.metadata for f in failing],
    environment__in=[f.environment for f in failing],
    build_id__lt=self.id,
).order_by("-build_id")[:limit]

# for each failing test, filter it's history locally, without going again to
# the database
for f in failing:
    f_history = [
        _t
        for _t in history
        if _t.metadata_id == f.metadata__in and _t.environment_id == f.environment_id
    ]
    f.set_confidence(self.project.build_confidence_threshold, f_history)
```

Reply:
```
).order_by("-build_id")[:limit]
```

If we set the limit here, we only get 20 tests in total. I need a history for each failing test for N builds. If I remove the limit, that would get all tests matching that.

Let me push something else to try here.

Reply:
yes, you are right. instead of limiting here you can do a query to get the ids the the last `limit` builds, and filter on those builds_ids instead.